README.md
[![Python](https://img.shields.io/badge/Python-3.5,3.6,3.7,3.8,3.9-blue.svg?style=flat-square)](/)
[![Django](https://img.shields.io/badge/Django-2.1,2.2,3.2,4.0-blue.svg?style=flat-square)](/)
[![License](https://img.shields.io/badge/License-BSD--3--Clause-blue.svg?style=flat-square)](/LICENSE)
[![PyPIv](https://img.shields.io/pypi/v/django-constrainedfilefield.svg?style=flat-square)](https://pypi.org/project/django-constrainedfilefield)
[![PyPIs](https://img.shields.io/pypi/status/django-constrainedfilefield.svg)](https://pypi.org/project/django-constrainedfilefield)
[![Build Status](https://travis-ci.org/mbourqui/django-constrainedfilefield.svg?branch=master)](https://travis-ci.org/mbourqui/django-constrainedfilefield)
[![Coverage Status](https://coveralls.io/repos/github/mbourqui/django-constrainedfilefield/badge.svg)](https://coveralls.io/github/mbourqui/django-constrainedfilefield)
[![Downloads](https://pepy.tech/badge/django-constrainedfilefield)](https://pepy.tech/project/django-constrainedfilefield)
<a href="https://github.com/psf/black"><img alt="Code style: black" src="https://img.shields.io/badge/code%20style-black-000000.svg"></a>

# ConstrainedFileField for Django

This Django app adds a new field type, `ConstrainedFileField`, that has the
capability of checking the file size and type. Also provides a javascript checker for the 
form field.


## Features
* File size limitation
* File type limitation
* Javascript file size checker


## Requirements
* [Python][] >= 3.5
* [Django]>= 2.1.15
* `python-magic` >= 0.4.2 *iff* you want to check the file type

## Installation

### Using PyPI
1. Run
   * `pip install django-constrainedfilefield`, or
   * `pip install django-constrainedfilefield[filetype]` to ensure `python-magic` is installed.
1. For windows, you must download the dll files and .magic file at https://github.com/pidydx/libmagicwin64 (32-bit version: http://gnuwin32.sourceforge.net/packages/file.htm)), add them to C:\\Windows\\System32 (or to a folder in your PATH), and set MAGIC_FILE_PATH="..." to the path of your .magic file in your settings.py. For more information about the files to download, go to: https://github.com/ahupp/python-magic/blob/43df08c5ed63d7aad839695f311ca1be2eeb1ecb/README.md#dependencies

### Using the source code
1. Make sure [Pandoc][] is installed
1. Run `./pypi_packager.sh`
1. Run `pip install dist/django_constrainedfilefield-x.y.z-[...].wheel`, where `x.y.z` must be replaced by the actual
   version number and `[...]` depends on your packaging configuration
1. For windows, you must download the dll files and .magic file at https://github.com/pidydx/libmagicwin64 (32-bit version: http://gnuwin32.sourceforge.net/packages/file.htm)), add them to C:\\Windows\\System32 (or to a folder in your PATH), and set MAGIC_FILE_PATH="..." to the path of your .magic file in your settings.py. For more information about the files to download, go to: https://github.com/ahupp/python-magic/blob/43df08c5ed63d7aad839695f311ca1be2eeb1ecb/README.md#dependencies

## Usage
### Validate single file
The field can be used in forms or model forms like a normal `FileField`. If a user tries to upload
a file which is too large or without a valid type, a form validation error will occur.

Note that the validation does not occur on the field itself (on `save()`), but when validated through a form.

#### Creating form from model
Create a model and add a field of type `ConstrainedFileField`. You can add a maximum size in bytes
and a list of valid mime types that will be allowed. The list of all mime types is available
here: http://www.iana.org/assignments/media-types/index.html.
Setting none of the above, it behaves like a regular `FileField`.
```
from django.db import models
from constrainedfilefield.fields import ConstrainedFileField

class TestModel(models.Model):
    the_file = ConstrainedFileField(
                            null=True,
                            blank=True,
                            upload_to='testfile',
                            content_types=['image/png'],
                            max_upload_size=10240
                                    )
```

```
from django import forms
from myproject.models import TestModel

class TestModelForm(forms.ModelForm):
    class Meta:
        model = TestModel
        fields = ['the_file']
```

#### Building a form
```
from django import forms
from constrainedfilefield.fields import ConstrainedFileField

class TestNoModelForm(forms.Form):
    the_file = ConstrainedFileField(
                            null=True,
                            blank=True,
                            upload_to='testfile',
                            content_types=['image/png'],
                            max_upload_size=10240
                                    ).formfield()
```

#### Javascript file size validation
Additionally, to prevent user uploading too large files, a javascript checker can be set to the 
form field. In order to achieve that, you need to

1. Add `constrainedfilefield` to the `INSTALLED_APPS`. This will load the
  javascripts from the static files.
1.  Activate this feature by setting `js_checker=True` when instantiating the
`ConstrainedFileField`.
1. Include the javascript in the template where the form field is used

    ```Django
    {% load static %}
    <script src="{% static 'constrainedfilefield/js/file_checker.js' %}"></script>
    ```


### Validate single image
Same as above, using `ConstrainedImageFileField` instead.

The `ConstrainedImageField` offers additional constraints:
* `[min|max]_upload_[width|height]` to define min/max dimensions, respectively width and height.


## Note on DOS attacks

Important note: the check of the file size is made by Django once the whole file has been uploaded
to the server and stored in a temp directory (or in memory if the file is small). Thus, this is
useful to guarantee the quota of the users, for example, but will not stop an attacking user that
wants to block the server by sending huge files (e. g. of several Gb).

To avoid this, you need to configure your front end to limit the size of uploaded files. How to do
it depends on the software you are using. For example, if you use apache, you should use
[**LimitRequestBody**](http://httpd.apache.org/docs/2.2/mod/core.html#limitrequestbody) directive.

This is a complementary measure, because you'll usually want normal users that exceed the size by a
reasonable amount to get a friendly form validation message, while attacking users will see how their
connection is abruptly cut before the file finishes uploading. So the recommended setting is to give
`max_upload_size` a small value (e.g. 5Mb) and `LimitRequestBody` a higher one (e.g. 100Mb).


## Credits

This is a fork of [django-validated-file](https://github.com/kaleidos/django-validated-file) from
[Kaleidos](https://github.com/kaleidos).

  [python]:     https://www.python.org/             "Python"
  [django]:     https://www.djangoproject.com/      "Django"
  [pandoc]:     http://pandoc.org/index.html        "Pandoc"

CHANGELOG.md
# Changelog

## 5.0.0 (02.09.2022)
### Added
- Support of Django 4.0

### Removed
- Support of Django 1.11

## 4.0.0 (23.05.2020)
### Added
- Support of Django 2.1
- Support of Django 2.2

### Removed
- Support of Django < 1.11
- Support of Python 2
- Support of Python 3.4

## 3.2.0 (03.05.2018)
* Support ImageField with constraints
* Update structure to have `fields` as package

## 3.1.0 (2017.12.24)
* Windows support for magic, thanks to PR #7 from @epicfaace

## 3.0.7 (2017.08.16)
* Package metadata updates
* Usage of lazy texts for translations

## 3.06 (2017.05.15)
* Close #4 (commit 982991c)
* Update messages
* Update tests (still ongoing)

## 3.0.4 (2017.05.11)
* Close #2 (commit 41ba77d)
* Update documentation
* Update tests (still ongoing)
* Integration with coveralls.io
* Fix js_checker issue in field deconstruct

## 3.0.3 (2017.04.12)
* Closes #1 (commit 1ff3fc8)
* Travis config
* Package metadata
* PyPI release

## 3.0.0 (2017.03.06)
* Update Field
  * New name
  * Drop quota
  * Add javascript file size validator
* Update tests
  * Now working with latest versions of Python and Django
  * Integration with Travis CI


constrainedfilefield/tests/models.py
from django.db import models

from constrainedfilefield.fields import ConstrainedFileField, ConstrainedImageField


class TestModel(models.Model):
    the_file = ConstrainedFileField(
        null=True,
        blank=True,
        upload_to="testfile",
        content_types=["image/png"],
        max_upload_size=10240,
    )


class TestDocModel(models.Model):
    the_file = ConstrainedFileField(
        null=False,
        blank=False,
        upload_to="testfile",
        content_types=[
            "application/msword",  # .doc
            "application/vnd.openxmlformats-officedocument.wordprocessingml.document",  # .docx
            "application/vnd.oasis.opendocument.text",  # .odt
        ],
        max_upload_size=10240,
    )


class TestImageModel(models.Model):
    the_image = ConstrainedImageField(
        null=True,
        blank=True,
        upload_to="testfile",
        content_types=['image/png'],)
    the_image_small = ConstrainedImageField(
        null=True,
        blank=True,
        upload_to="testfile",
        content_types=["image/png"],
        min_upload_size=1024,
        max_upload_size=10240,
        min_upload_height=100,
        max_upload_height=500,
        min_upload_width=100,
        max_upload_width=500,
    )
    the_image_large = ConstrainedImageField(
        null=True,
        blank=True,
        upload_to="testfile",
        content_types=["image/png"],
        min_upload_size=10240,
        max_upload_size=20480,
        min_upload_height=3000,
        max_upload_height=5000,
        min_upload_width=3000,
        max_upload_width=5000,
    )


class TestModelJs(models.Model):
    the_file = ConstrainedFileField(
        null=True,
        blank=True,
        upload_to="testfile",
        content_types=["image/png"],
        max_upload_size=10240,
        js_checker=True,
    )
    the_image = ConstrainedImageField(
        null=True,
        blank=True,
        upload_to="testfile",
        content_types=["image/png"],
        min_upload_size=1024,
        max_upload_size=10240,
        js_checker=True,
    )


class TestModelNoValidate(models.Model):
    the_file = ConstrainedFileField(null=True, blank=True, upload_to="testfile")
    the_image = ConstrainedImageField(null=True, blank=True, upload_to="testfile")


class TestContainer(models.Model):
    name = models.CharField(max_length=100)


class TestElement(models.Model):
    container = models.ForeignKey(
        TestContainer, on_delete=models.CASCADE, related_name="test_elements"
    )
    the_file = ConstrainedFileField(
        null=True,
        blank=True,
        upload_to="testfile",
        content_types=["image/png", "image/jpeg"],
    )


constrainedfilefield/tests/tests/backend.py
# -*- coding: utf-8 -*-
import os.path

from django.conf import settings
from django.core.files import File
from django.core.files.uploadedfile import SimpleUploadedFile
from django.test import TestCase

from constrainedfilefield.tests.forms import (
    TestModelForm,
    TestDocModelForm,
    TestImageModelForm,
    TestModelFormJs,
    TestModelNoValidateForm,
    TestElementForm,
    TestNoModelForm,
    TestNoModelJsForm,
)
from constrainedfilefield.tests.models import TestModel, TestImageModel, TestContainer


class ConstrainedFileFieldTest(TestCase):
    SAMPLE_FILES_PATH = os.path.join(settings.BASE_DIR, "sample_files")

    def test_create_empty_instance(self):
        TestModel.objects.create()

    def test_create_instance_with_file(self):
        instance = TestModel.objects.create(
            the_file=File(self._get_sample_file("image2k.png"), "the_file.png")
        )

        self._check_file_url(instance.the_file, "the_file.png")

        instance.the_file.delete()
        instance.delete()

    def test_create_instance_with_images(self):
        instance = TestImageModel.objects.create(
            the_image=File(self._get_sample_file("image2k.png"), "the_image.png"),
            the_image_small=File(
                self._get_sample_file("image2k.png"), "the_image_small.png"
            ),
            the_image_large=File(
                self._get_sample_file("image15k.png"), "the_image_large.png"
            ),
        )

        self._check_file_url(instance.the_image, "the_image.png")
        self._check_file_url(instance.the_image_small, "the_image_small.png")
        self._check_file_url(instance.the_image_large, "the_image_large.png")

        instance.the_image.delete()
        instance.the_image_small.delete()
        instance.the_image_large.delete()
        instance.delete()

    def test_form_ok(self):
        form = self._create_bound_test_model_form(
            form_class=TestModelForm,
            orig_filename="image2k.png",
            dest_filename="the_file.png",
            content_type="image/png",
        )
        self.assertTrue(form.is_valid())
        instance = form.save()

        self._check_file_url(instance.the_file, "the_file.png")

        instance.the_file.delete()
        instance.delete()

    def test_form_doc_ok(self):
        # ODT
        form = self._create_bound_test_model_form(
            form_class=TestDocModelForm,
            orig_filename="document15k.odt",
            dest_filename="the_file.odt",
            content_type="application/vnd.oasis.opendocument.text",
        )
        self.assertTrue(form.is_valid())
        instance = form.save()
        self._check_file_url(instance.the_file, "the_file.odt")
        instance.the_file.delete()
        instance.delete()

        # DOC
        form = self._create_bound_test_model_form(
            form_class=TestDocModelForm,
            orig_filename="document15k.doc",
            dest_filename="the_file.doc",
            content_type="application/msword",
        )
        self.assertTrue(form.is_valid())
        instance = form.save()
        self._check_file_url(instance.the_file, "the_file.doc")
        instance.the_file.delete()
        instance.delete()

        # DOCX
        # Requires a native .docx file
        # form = self._create_bound_test_model_form(form_class=TestModelDocForm,
        #                                           orig_filename='document15k.docx',
        #                                           dest_filename='the_file.docx',
        #                                           content_type='application/vnd.openxmlformats-officedocument.wordprocessingml.document')
        # self.assertTrue(form.is_valid())
        # instance = form.save()
        # self._check_file_url(instance.the_file, 'the_file.docx')
        # instance.the_file.delete()
        # instance.delete()

    def test_form_js_ok(self):
        form = self._create_bound_test_model_form(
            form_class=TestModelFormJs,
            orig_filename="image2k.png",
            dest_filename="the_file.png",
            content_type="image/png",
        )
        self.assertTrue(form.is_valid())
        instance = form.save()

        self._check_file_url(instance.the_file, "the_file.png")

        instance.the_file.delete()
        instance.delete()

    def test_form_invalid_size(self):
        form = self._create_bound_test_model_form(
            form_class=TestModelForm,
            orig_filename="image15k.png",
            dest_filename="the_file.png",
            content_type="image/png",
        )
        self.assertFalse(form.is_valid())
        self.assertEqual(len(form.errors), 1)
        self.assertEqual(len(form.errors["the_file"]), 1)
        self.assertEqual(
            form.errors["the_file"][0],
            u"File size exceeds limit: 14.2 KB. Limit is 10.0 KB.",
        )

    def test_form_valid_image_size(self):
        files = {
            "the_image": self._create_simple_uploaded_file(
                orig_filename="image2k.png",
                dest_filename="the_image.png",
                content_type="image/png",
            ),
            "the_image_small": self._create_simple_uploaded_file(
                orig_filename="image2k.png",
                dest_filename="the_image_small.png",
                content_type="image/png",
            ),
            "the_image_large": self._create_simple_uploaded_file(
                orig_filename="image15k.png",
                dest_filename="the_image_large.png",
                content_type="image/png",
            ),
        }
        form = TestImageModelForm(data={}, files=files)
        self.assertTrue(form.is_valid())

    def test_form_invalid_image_size(self):
        files = {
            "the_image": self._create_simple_uploaded_file(
                orig_filename="image2k.png",
                dest_filename="the_image.png",
                content_type="image/png",
            ),
            "the_image_small": self._create_simple_uploaded_file(
                orig_filename="image15k.png",
                dest_filename="the_image_small.png",
                content_type="image/png",
            ),
            "the_image_large": self._create_simple_uploaded_file(
                orig_filename="image2k.png",
                dest_filename="the_image_large.png",
                content_type="image/png",
            ),
        }
        form = TestImageModelForm(data={}, files=files)
        self.assertFalse(form.is_valid())
        self.assertEqual(len(form.errors), 2)
        self.assertEqual(len(form.errors["the_image_small"]), 3)
        self.assertEqual(
            form.errors["the_image_small"][0],
            u"File size exceeds limit: 14.2 KB. Limit is 10.0 KB.",
        )
        self.assertEqual(
            form.errors["the_image_small"][1],
            u"File height exceeds limit: 14586. Limit is 500.",
        )
        self.assertEqual(
            form.errors["the_image_small"][2],
            u"File width exceeds limit: 14586. Limit is 500.",
        )
        self.assertEqual(len(form.errors["the_image_large"]), 3)
        self.assertEqual(
            form.errors["the_image_large"][0],
            u"File size below limit: 2.1 KB. Limit is 10.0 KB.",
        )
        self.assertEqual(
            form.errors["the_image_large"][1],
            u"File height below limit: 2120. Limit is 3000.",
        )
        self.assertEqual(
            form.errors["the_image_large"][2],
            u"File width below limit: 2120. Limit is 3000.",
        )

    def test_form_invalid_filetype(self):
        form = self._create_bound_test_model_form(
            form_class=TestModelForm,
            orig_filename="document1k.pdf",
            dest_filename="the_file.pdf",
            content_type="application/pdf",
        )
        self.assertFalse(form.is_valid())
        self.assertEqual(len(form.errors), 1)
        self.assertEqual(len(form.errors["the_file"]), 1)
        self.assertEqual(
            form.errors["the_file"][0],
            u"Unsupported file type: application/pdf. "
            u"Allowed types are ['image/png'].",
        )

    def test_form_invalid_image_filetype(self):
        files = {
            "the_image": self._create_simple_uploaded_file(
                orig_filename="unsupported_image.jpg",
                dest_filename="the_file.jpg",
                content_type="image/jpeg",
            ),
        }
        form = TestImageModelForm(data={}, files=files)
        self.assertFalse(form.is_valid())
        self.assertEqual(len(form.errors), 1)
        self.assertEqual(len(form.errors["the_image"]), 1)
        self.assertEqual(
            form.errors["the_image"][0],
            u"Unsupported file type: image/jpeg. "
            u"Allowed types are ['image/png']."
        )

    def test_form_invalid_filetype_and_size(self):
        form = self._create_bound_test_model_form(
            form_class=TestModelForm,
            orig_filename="document15k.pdf",
            dest_filename="the_file.pdf",
            content_type="application/pdf",
        )
        self.assertFalse(form.is_valid())
        self.assertEqual(len(form.errors), 1)
        self.assertEqual(len(form.errors["the_file"]), 1)
        self.assertEqual(
            form.errors["the_file"][0],
            u"File size exceeds limit: 14.9 KB. Limit is 10.0 KB.",
        )

    def test_form_fake_filetype(self):
        form = self._create_bound_test_model_form(
            form_class=TestModelForm,
            orig_filename="document1k.pdf",
            dest_filename="the_file.pdf",
            content_type="image/png",
        )
        self.assertFalse(form.is_valid())
        self.assertEqual(len(form.errors), 1)
        self.assertEqual(len(form.errors["the_file"]), 1)
        self.assertEqual(
            form.errors["the_file"][0],
            u"Unsupported file type: application/pdf. "
            u"Allowed types are ['image/png'].",
        )

    def test_form_no_validate(self):
        form = self._create_bound_test_model_form(
            form_class=TestModelNoValidateForm,
            orig_filename="document15k.pdf",
            dest_filename="the_file.pdf",
            content_type="application/pdf",
        )
        self.assertTrue(form.is_valid())
        instance = form.save()

        self._check_file_url(instance.the_file, "the_file.pdf")

        instance.the_file.delete()
        instance.delete()

    def test_form_null_file(self):
        form = self._create_bound_test_model_form(form_class=TestModelNoValidateForm)
        self.assertTrue(form.is_valid())
        instance = form.save()

        self.assertEqual(instance.the_file, None)

        instance.delete()

    def test_nomodel_form_ok(self):
        form = self._create_bound_test_model_form(
            form_class=TestNoModelForm,
            orig_filename="image2k.png",
            dest_filename="the_file.png",
            content_type="image/png",
        )
        self.assertTrue(form.is_valid())

        for uploaded_file in form.files.values():
            uploaded_file.close()

    def test_nomodel_form_js_ok(self):
        form = self._create_bound_test_model_form(
            form_class=TestNoModelJsForm,
            orig_filename="image2k.png",
            dest_filename="the_file.png",
            content_type="image/png",
        )
        self.assertTrue(form.is_valid())

        for uploaded_file in form.files.values():
            uploaded_file.close()

    # -------
    # Utilities

    def _get_sample_file(self, filename):
        path = os.path.join(self.SAMPLE_FILES_PATH, filename)
        return open(path, mode="rb")

    def _check_file_url(self, filefield, filename):
        url = settings.MEDIA_URL + filefield.field.upload_to + "/" + filename
        self.assertEqual(filefield.url, url)

    def _create_simple_uploaded_file(self, orig_filename, dest_filename, content_type):
        return SimpleUploadedFile(
            name=dest_filename,
            content=self._get_sample_file(orig_filename).read(),
            content_type=content_type,
        )

    def _create_bound_test_model_form(
        self, form_class, orig_filename=None, dest_filename=None, content_type=None
    ):
        if orig_filename and dest_filename and content_type:
            uploaded_file = self._create_simple_uploaded_file(
                orig_filename, dest_filename, content_type=content_type,
            )
            files = {"the_file": uploaded_file}
        else:
            files = {}
        form = form_class(data={}, files=files)
        return form

    def _create_container(self, name):
        return TestContainer.objects.create(name=name)

    def _add_element(self, container, orig_filename, dest_filename):
        return container.test_elements.create(
            the_file=File(self._get_sample_file(orig_filename), dest_filename)
        )

    def _create_unbound_test_element_form(self, container):
        return TestElementForm(container=container)

    def _create_bound_test_element_form(
        self, container, orig_filename=None, dest_filename=None, content_type=None
    ):
        if orig_filename and dest_filename and content_type:
            uploaded_file = self._create_simple_uploaded_file(
                orig_filename, dest_filename, content_type=content_type,
            )
            files = {"the_file": uploaded_file}
        else:
            files = {}
        form = TestElementForm(container=container, data={}, files=files)
        return form


constrainedfilefield/fields/file.py
# -*- coding: utf-8 -*-
__all__ = ["ConstrainedFileField"]

import os

from django import forms
from django.conf import settings
from django.db import models
from django.template.defaultfilters import filesizeformat
from django.utils.translation import ugettext_lazy as _


class ConstrainedFileField(models.FileField):
    """
    A FileField with additional constraints. Namely, the file size and type can be restricted. If
    using the types, the magic library is required. Setting neither a file size nor type behaves
    like a regular FileField.

    Parameters
    ----------
    content_types : list of str
        List containing allowed content_types. Example: ['application/pdf', 'image/jpeg']
    max_upload_size : int
        Maximum file size allowed for upload, in bytes
            1 MB - 1048576 B - 1024**2 B - 2**20 B
            2.5 MB - 2621440 B
            5 MB - 5242880 B
            10 MB - 10485760 B
            20 MB - 20971520 B
            33 MiB - 2**25 B
            50 MB - 5242880 B
            100 MB 104857600 B
            250 MB - 214958080 B
            500 MB - 429916160 B
            1 GiB - 1024 MiB - 2**30 B
    js_checker : bool
        Add a javascript file size checker to the form field
    mime_lookup_length : int

    See Also
    --------
    Based on https://github.com/kaleidos/django-validated-file/blob/master/validatedfile/fields.py
    With inspiration from http://stackoverflow.com/a/9016664

    """

    description = _("A file field with constraints on size and/or type")

    def __init__(self, *args, **kwargs):
        self.max_upload_size = kwargs.pop("max_upload_size", 0)
        assert isinstance(self.max_upload_size, int) and self.max_upload_size >= 0
        self.content_types = kwargs.pop("content_types", [])
        self.mime_lookup_length = kwargs.pop("mime_lookup_length", 4096)
        assert isinstance(self.mime_lookup_length, int) and self.mime_lookup_length >= 0
        self.js_checker = kwargs.pop("js_checker", False)

        super(ConstrainedFileField, self).__init__(*args, **kwargs)

    def clean(self, *args, **kwargs):
        data = super(ConstrainedFileField, self).clean(*args, **kwargs)

        if self.max_upload_size and data.size > self.max_upload_size:
            # Ensure no one bypasses the js checker
            raise forms.ValidationError(
                _("File size exceeds limit: %(current_size)s. Limit is %(max_size)s.")
                % {
                    "max_size": filesizeformat(self.max_upload_size),
                    "current_size": filesizeformat(data.size),
                }
            )

        if self.content_types:
            import magic

            file = data.file
            uploaded_content_type = getattr(file, "content_type", "")

            # magic_file_path used only for Windows.
            magic_file_path = getattr(settings, "MAGIC_FILE_PATH", None)
            if magic_file_path and os.name == "nt":
                mg = magic.Magic(mime=True, magic_file=magic_file_path)
            else:
                mg = magic.Magic(mime=True)
            content_type_magic = mg.from_buffer(file.read(self.mime_lookup_length))
            file.seek(0)

            # Prefer mime-type from magic over mime-type from http header
            if uploaded_content_type != content_type_magic:
                uploaded_content_type = content_type_magic

            if uploaded_content_type not in self.content_types:
                raise forms.ValidationError(
                    _("Unsupported file type: %(type)s. Allowed types are %(allowed)s.")
                    % {"type": content_type_magic, "allowed": self.content_types}
                )
        return data

    def formfield(self, **kwargs):
        """
        Usual Form for a django.models.FileField with optional javascript file size
        checker. Can thus be customized as any other Form for a django.models.FileField.

        Parameters
        ----------
        kwargs
            will be passed to super().formfield()

        Returns
        -------
        django.forms.FileField

        """
        formfield = super(ConstrainedFileField, self).formfield(**kwargs)
        if self.js_checker:
            formfield.widget.attrs.update(
                {"onchange": "validateFileSize(this, 0, %d);" % (self.max_upload_size,)}
            )
        return formfield

    def deconstruct(self):
        name, path, args, kwargs = super(ConstrainedFileField, self).deconstruct()
        if self.max_upload_size:
            kwargs["max_upload_size"] = self.max_upload_size
        if self.content_types:
            kwargs["content_types"] = self.content_types
        if self.mime_lookup_length:
            kwargs["mime_lookup_length"] = self.mime_lookup_length
        if self.js_checker:
            kwargs["js_checker"] = self.js_checker
        return name, path, args, kwargs

    def __str__(self):
        if hasattr(self, "model"):
            return super(ConstrainedFileField).__str__()
        else:
            return self.__class__.__name__
